futures-util/src/future/shared.rs

```diff
@@ -156,153 +196,76 @@ impl<F> Future for Shared<F>
+                if let FutureOrOutput::Future(future) = &mut *this.inner.future_or_output.get() {
+                    PinMut::new_unchecked(future).poll(&mut cx)
+                } else {
+                    unreachable!()
+                }
             };
             match res {
-                Ok(Async::Pending) => {
+                Poll::Pending => {
                     // Not ready, try to release the handle
-                    match self.inner.notifier.state.compare_and_swap(POLLING, IDLE, SeqCst) {
+                    match this.inner.notifier.state.compare_and_swap(POLLING, IDLE, SeqCst) {
                         POLLING => {
                             // Success
-                            return Ok(Async::Pending);
+                            return Poll::Pending;
                         }
                         REPOLL => {
                             // Gotta poll again!
-                            let prev = self.inner.notifier.state.swap(POLLING, SeqCst);
+                            let prev = this.inner.notifier.state.swap(POLLING, SeqCst);
                             assert_eq!(prev, REPOLL);
                         }
                         _ => unreachable!(),
                     }
-
-                }
-                Ok(Async::Ready(i)) => {
-                    unsafe {
-                        (*self.inner.result.get()) = Some(Ok(SharedItem { item: Arc::new(i) }));
-                    }
-
-                    break;
                 }
-                Err(e) => {
+                Poll::Ready(output) => {
+                    let output = Arc::new(output);
                     unsafe {
-                        (*self.inner.result.get()) = Some(Err(SharedError { error: Arc::new(e) }));
+                        *this.inner.future_or_output.get() =
+                            FutureOrOutput::Output(output.clone());
                     }
-
-                    break;
+                    // Complete the future
+                    this.inner.notifier.state.store(COMPLETE, SeqCst);
+                    Wake::wake(&this.inner.notifier);
+                    return Poll::Ready(output);
                 }
             }
         }
-
-        self.complete();
-        unsafe { self.clone_result().map(Async::Ready) }
     }
 }
 
 impl<F> Clone for Shared<F> where F: Future {
     fn clone(&self) -> Self {
-        let next_clone_id = self.inner.next_clone_id.fetch_add(1, SeqCst);
-
         Shared {
             inner: self.inner.clone(),
-            waiter: next_clone_id,
+            waker_key: NULL_WAKER_KEY,
         }
     }
 }
 
 impl<F> Drop for Shared<F> where F: Future {
     fn drop(&mut self) {
-        let mut waiters = self.inner.notifier.waiters.lock().unwrap();
-        waiters.remove(&self.waiter);
+        if self.waker_key != NULL_WAKER_KEY {
+            if let Ok(mut wakers) = self.inner.notifier.wakers.lock() {
+                mem::replace(&mut *wakers, Slab::new());
```

Comment:
I think this is a bug. You don't want to remove all the wakers. Just the one for this `Shared`

Reply:
Haha yup-- I was just fixing that. Where did you imagine the `Option` being set to `None`?

Reply:
I think it's probably fine to keep the Slab around-- shrinking it before all of the children have received the value and are ready to drop their `Shared` seems unnecessary.